gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/SchedulerLeaseAlgoHandler.java

```diff
@@ -0,0 +1,179 @@
+        break;
+      case "LeasedToAnotherStatus":
+        scheduleReminderForEvent(jobProps, (LeasedToAnotherStatus) leaseAttemptStatus, flowAction, eventTimeMillis);
+        break;
+      case "NoLongerLeasingStatus":
+        break;
+      default:
+    }
+  }
+
+  // Called after obtaining a lease to persist the flow action to {@link DagActionStore} and mark the lease as done
+  private boolean finalizeLease(LeaseObtainedStatus status, DagActionStore.DagAction flowAction) {
+    try {
+      this.dagActionStore.addDagAction(flowAction.getFlowGroup(), flowAction.getFlowName(),
+          flowAction.getFlowExecutionId(), flowAction.getFlowActionType());
+      if (this.dagActionStore.exists(flowAction.getFlowGroup(), flowAction.getFlowName(),
+          flowAction.getFlowExecutionId(), flowAction.getFlowActionType())) {
+        // If the flow action has been persisted to the {@link DagActionStore} we can close the lease
+        this.numLeasesCompleted.mark();
+        return this.multiActiveLeaseArbiter.completeLeaseUse(flowAction, status.getEventTimestamp(),
+            status.getMyLeaseAcquisitionTimestamp());
+      }
+    } catch (IOException | SQLException e) {
+      throw new RuntimeException(e);
+    }
+    // TODO: should this return an error or print a warning log if failed to commit to dag action store?
+    return false;
+  }
+
+  /**
+   * This method is used by {@link SchedulerLeaseAlgoHandler.handleNewSchedulerEvent} to schedule a reminder for itself
+   * to check on the other participant's progress to finish acting on a flow action after the time the lease should
+   * expire.
+   * @param jobProps
+   * @param status used to extract event to be reminded for and the minimum time after which reminder should occur
+   * @param originalEventTimeMillis the event timestamp we were originally handling
+   * @param flowAction
+   */
+  private void scheduleReminderForEvent(Properties jobProps, LeasedToAnotherStatus status,
+      DagActionStore.DagAction flowAction, long originalEventTimeMillis) {
+    // Add a small randomization to the minimum reminder wait time to avoid 'thundering herd' issue
+    String cronExpression = createCronFromDelayPeriod(status.getMinimumReminderWaitMillis() + random.nextInt(staggerUpperBoundSec));
+    jobProps.setProperty(ConfigurationKeys.JOB_SCHEDULE_KEY, cronExpression);
+    // Ensure we save the event timestamp that we're setting reminder for, in addition to our own event timestamp which may be different
+    jobProps.setProperty(ConfigurationKeys.SCHEDULER_REMINDER_EVENT_TIMESTAMP_MILLIS_KEY, String.valueOf(status.getReminderEventTimeMillis()));
+    jobProps.setProperty(ConfigurationKeys.SCHEDULER_NEW_EVENT_TIMESTAMP_MILLIS_KEY, String.valueOf(status.getReminderEventTimeMillis()));
+    JobKey key = new JobKey(flowAction.getFlowName(), flowAction.getFlowGroup());
+    Trigger trigger = this.jobScheduler.getTrigger(key, jobProps);
```

Comment:
sorry, I guess I'm unfamiliar: what's this `Trigger` we get from one scheduler and give to another?

Reply:
I added some comments to clarify but there is an existing function in `JobScheduler` named `getTrigger` (will rename this to `createTrigger`) we are using to create a new Trigger for the job that will fire after the lease should expire and passing it to the `SchedulerService` to schedule it.